Applications/Spire/Source/KeyBindings/CancelKeyBindingsTableModel.cpp

```diff
@@ -0,0 +1,171 @@
+#include "Spire/KeyBindings/CancelKeyBindingsTableModel.hpp"
+
+using namespace Spire;
+
+namespace {
+  const auto ROW_COUNT = 13;
+  const auto COLUMN_COUNT = 2;
+
+  auto get_action(int index) {
+    static auto actions = std::vector<KeyBindings::CancelAction>({
```

Comment:
I don't see a benefit of using this compared to a simple `switch` statement.  Constructing `std::vector` requires heap allocation, nor does it make the intent any clearer.

Reply:
My issue with the enum is that if the an 'incorrect' value is passed to get_action(int), I would prefer that it crashes. Is there a suitable default for the case where an invalid index is passed to this function?

Reply:
There is no guarantee that the code as is will crash either.
You should either make handling incorrect values (likely in the public `get_action` method, not here) robust - e.g. by returning an optional, not undefined behavior - or not handle those at all. The second approach is better here as users have a way to identify which rows are correct and which are not.

Reply:
For the second approach should I add a NONE action to the CancelAction enum?

Reply:
No. Basically, the behavior of the static function `get_action` is defined for valid indexes, that is 0 <= i <= ROW_COUNT. Passing an index outside of the boundaries is an error, and program behavior is undefined should that happen.
Similarly, the behavior of method `CancelKeyBindingsTableModel::get_action` is only defined for valid indexes. There is no need to define it for invalid indexes as index validity can be checked by a users using `rowCount` method.

Reply:
So in practice for invalid indexes it makes no difference what it returns because it's undefined behavior?

Reply:
Not only what it returns, but also whether it returns anything and whether the rest of the program will run normally.
You have an option to handle invalid inputs using explicit checks, but out-of-bound array accesses achieve little as that is, by standard, undefined behavior on its own. There are very few ways to crash a C++ program in a predictable way (one involves unhandled exceptions in worker threads, which will result in `std::terminate`) except by explicit use of system calls or their wrappers.
Regardless of the above, as a convention for, as far as I know, all our projects, we do not handle any invalid inputs except for those coming directly from a user or an unreliable source. There will be some place where a user (implicitly) passes a row number but, as a rule of thumb, input validation should be done as early as possible. As long as a caller of your code has means to validate the arguments they are going to pass, it is their responsibility (or their users') to do so.

Reply:
Would it be better to return an optional as the default condition then?

Reply:
No.